include/aws/common/linked_list.h

```diff
@@ -37,28 +37,115 @@ struct aws_linked_list {
+#if (AWS_DEEP_CHECKS == 1)
+        return aws_linked_list_is_valid_deep(list);
+#else
+        return true;
+#endif
+    }
+    return false;
+}
+
+/**
+ * Checks that the prev of the next pointer of a node points to the
+ * node. As this checks whether the [next] connection of a node is
+ * bidirectional, it returns false if used for the list tail.
+ */
+AWS_STATIC_IMPL bool aws_linked_list_node_next_is_valid(const struct aws_linked_list_node *node) {
+    return node && node->next && node->next->prev == node;
+}
+
+/**
+ * Checks that the next of the prev pointer of a node points to the
+ * node. Similarly to the above, this returns false if used for the
+ * head of a list.
+ */
+AWS_STATIC_IMPL bool aws_linked_list_node_prev_is_valid(const struct aws_linked_list_node *node) {
+    return node && node->prev && node->prev->next == node;
+}
+
+/**
+ * Checks that a linked list satisfies double linked list connectivity
+ * constraints. This check is O(n) as it traverses the whole linked
+ * list to ensure that tail is reachable from head (and vice versa)
+ * and that every connection is bidirectional.
+ *
+ * Note: This check could go into an inginite loop if the list is
+ * circular.
+ */
+AWS_STATIC_IMPL bool aws_linked_list_is_valid_deep(const struct aws_linked_list *list) {
+    if (!list) {
+        return false;
+    }
+    /* This could go into an infinite loop for a circular list */
+    const struct aws_linked_list_node *temp = &list->head;
+    /* Head must reach tail by following next pointers */
+    bool head_reaches_tail = false;
+    /* Next and prev pointers should connect the same nodes */
+    bool are_links_valid = true;
+    /* By satisfying both of the above, we also guarantee that tail
+     * reaches head by following prev pointers */
+    while (temp) {
+        if (temp == &list->tail) {
+            head_reaches_tail = true;
+            break;
+        } else if (!aws_linked_list_node_next_is_valid(temp)) {
+            are_links_valid = false;
+            break;
+        }
+        temp = temp->next;
+    }
+    return head_reaches_tail && are_links_valid;
+}
+
 /**
  * Initializes the list. List will be empty after this call.
  */
 AWS_STATIC_IMPL void aws_linked_list_init(struct aws_linked_list *list) {
+    AWS_PRECONDITION(list);
     list->head.next = &list->tail;
     list->head.prev = NULL;
     list->tail.prev = &list->head;
     list->tail.next = NULL;
+    AWS_POSTCONDITION(aws_linked_list_is_valid(list));
+    AWS_POSTCONDITION(aws_linked_list_empty(list));
 }
 
 /**
  * Returns an iteration pointer for the first element in the list.
  */
 AWS_STATIC_IMPL struct aws_linked_list_node *aws_linked_list_begin(const struct aws_linked_list *list) {
-    return list->head.next;
+    AWS_PRECONDITION(aws_linked_list_is_valid(list));
+    struct aws_linked_list_node *rval = list->head.next;
+    AWS_POSTCONDITION(aws_linked_list_is_valid(list));
```

Comment:
Could add some of the checks from the proof as additional postconditions?

Reply:
Done